### ISSUE TITLE
Infra: sequence replay buffers can sample a window that is not a trajectory, and cannot say when they have none

### ISSUE THREAD

**Issue by le2333** [le2333/RTRRL#57]
The two sequence replay buffers in `memo/memorax/buffers/` share three defects
that predate #43 and are not that issue's to fix. They are filed together
because they are the same class of problem — the sampler answers a question it
cannot actually answer and does not say so — and separately from #43 because
fixing them changes what DQN and R2D2 draw, which has to be judged on those
learners' terms.

Filing this is not a claim that either learner is currently producing wrong
numbers. Two of the three need a specific condition to bite, and whether that
condition is reached in the runs we have actually launched is part of what this
issue is asking.

### 1. `episode_buffer.py` falls back to position zero when nothing is admissible

`sample_fn` computes a start mask, and when it is empty:

```python
def _fallback_beginning(key):
    rows = jax.random.randint(key, (sample_batch_size,), 0, add_batch_size)
    starts = jnp.zeros((sample_batch_size,), dtype=jnp.int32)
    return rows, starts

rows, starts = jax.lax.cond(total > 0, _sample_from_starts, _fallback_beginning, rng_key)
```

So a buffer with no legal start returns the same window from the head of each
row, forever, and the learner trains on it and reports a loss. Under JIT there
is nothing to raise, which is why the fallback exists — but the answer is not
to invent a window. It is for `can_sample` to be false.

The disconnect is that `can_sample` reports on the buffer's *length* while
`sample` needs an *eligible start*, and those are different questions whenever
the start rule is anything but "every position". `make_episode_buffer` does not
give `can_sample` the start rule at all.

Note that `prioritised_episode_buffer.py` already gets this right:

```python
eligible = start_flags & positions & (priorities > 0)
return length_ready & jnp.any(eligible)
```

so the design exists in this repository; the non-prioritised buffer just does
not use it. The fix is to give `make_episode_buffer` the same treatment and
then delete the fallback.

### 2. Both buffers admit every position once the ring is full

```python
def _full() -> Array:
    return jnp.ones((max_length_time_axis,), dtype=bool)
```

in `episode_buffer.py`, and the identical line in
`prioritised_episode_buffer.py`. Combined with the modulo gather:

```python
time_idx = (starts[:, None] + jnp.arange(sample_sequence_length)) % max_length_time_axis
```

a window starting within `sample_sequence_length - 1` of the write head reads
across the seam: the rows before the head are the newest transitions and the
rows at it are the oldest, and they never followed one another. The sample is a
splice of two moments presented to the learner as a trajectory.

For a start rule keyed on `done`, this is bounded by how often a spliced window
happens to begin at a genuine episode start — but "bounded" is not "absent",
and the window it produces is not a trajectory at all. It should be excluded,
not relied on to be rare.

### 3. `episode_buffer.py` draws its minibatch with replacement

```python
rows = jax.random.choice(..., replace=True)
starts = jax.vmap(lambda k, p: jax.random.choice(k, ..., replace=True))(...)
```

This one is **probably fine for R2D2 and DQN and is listed for completeness**,
not as a defect to fix by default. Where the unit of replay is a transition or
a fixed-length sequence, drawing items independently is the standard reading of
uniform replay, and prioritised replay samples items independently by
definition. It is worth a deliberate decision recorded in the code rather than
an accident of which `jax.random.choice` flag was passed.

### Why #43 does not fix these

#43 needed a sampler whose unit is the *episode*, drawn without replacement,
which is not something a per-position start rule can express regardless of how
these bugs are resolved. It therefore ships its own
`make_uniform_episode_window_buffer` (transition ring plus an episode index in
logical time), which sidesteps all three: the seam cannot be crossed because an
episode is entirely stored exactly when its first transition is, and
`can_sample` counts eligible episodes. `memo/docs/drqn-reproduction.md` has the
reasoning under "Why replay has its own buffer".

Nothing in that work touches these two buffers, and a `seam_margin` parameter
added to `episode_buffer.py` mid-branch was reverted for the same reason.

### Scope

- [ ] Determine whether any launched DQN or R2D2 run has actually met the
      conditions for (1) or (2), and say so on this issue before changing
      anything. If a run has, its numbers need re-examining.
- [ ] `make_episode_buffer`: make `can_sample` eligibility-aware, mirroring the
      prioritised buffer, and remove `_fallback_beginning`.
- [ ] Both buffers: exclude the starts whose window would span the write head
      once the ring is full.
- [ ] Record the with-replacement decision in the code, either way.
- [ ] Regression tests naming which positions are admitted, rather than
      asserting that splices are uncommon.

Low priority relative to R1's own arms, but it should not stay unwritten.
